### PR TITLE
[Multiple Datasource] Use data source filter function before rendering

### PR DESCRIPTION
**Pull request by BionIT**
### Description
This change will store all data sources fetched during initial load, then apply the datasource filter function before rendering so that it can filter based on changes in the filter function.

### Issues Resolved

Fixes https://github.com/opensearch-project/OpenSearch-Dashboards/issues/6174

## Screenshot

https://github.com/opensearch-project/OpenSearch-Dashboards/assets/17714150/27b5e077-3a90-4094-a311-19c402de1dd3

## Testing the changes
The following steps were performed in the recording:
1. go to add sample data, and data source selector should render correctly with currently available data sources
2. add sample data from a data source is successful, and dashboard renders successfully
3. go to dev tool, and data source selector should render correctly with currently available data sources
4. make a query and is successfully returning response
5. go to search relevance and the data source selector should render correctly, with local cluster as default option and no other options available to choose
6. select a data source from the data source menu in the nav bar which acts as filter, then check the data source selector, which should have the selected data source as another option in addition to local cluster

### Check List

- [ ] All tests pass
  - [ ] `yarn test:jest`
  - [ ] `yarn test:jest_integration`
- [ ] New functionality includes testing.
- [ ] New functionality has been documented.
- [ ] Update [CHANGELOG.md](./../CHANGELOG.md)
- [ ] Commits are signed per the DCO using --signoff

### CODE REVIEW

**Review by ruanyl — APPROVED** [review]

**Review by xinruiba — APPROVED** [review]

**Review by ZilongX — APPROVED** [review]

**Comment by ZilongX** [comment]
LGTM, picked up latest main branch and now rerun all checks

File: src/plugins/data_source_management/public/components/data_source_selector/data_source_selector.tsx [hunk 1]
```diff
@@ -119,6 +99,16 @@ export class DataSourceSelector extends React.Component<
       this.props.placeholderText === undefined
         ? 'Select a data source'
         : this.props.placeholderText;
+
+    const dataSources = this.props.dataSourceFilter
```
Comment (Flyingliuhub) [inline src/plugins/data_source_management/public/components/data_source_selector/data_source_selector.tsx:103]: Do we have UT to cover this changes? Thanks

**Review by bandinib-amzn — APPROVED** [review]

**Review by Flyingliuhub — APPROVED** [review]

File: src/plugins/data_source_management/public/components/data_source_selector/data_source_selector.tsx (hunk 1 above)
Comment (bandinib-amzn) [reply src/plugins/data_source_management/public/components/data_source_selector/data_source_selector.tsx:103]: +1. Approved but I'm wondering if we need UT for this change.

Comment (BionIT) [reply src/plugins/data_source_management/public/components/data_source_selector/data_source_selector.tsx:103]: yes, we use the same unit test for the filtering which was added when the datasource filter was introduced https://github.com/opensearch-project/OpenSearch-Dashboards/blob/main/src/plugins/data_source_management/public/components/data_source_selector/data_source_selector.test.tsx#L137

**Pull request merged**